### PR TITLE
Do not scan nodes preceding formatted region, just skip over them

### LINKED ISSUE #7686: Crash when formatting template strings
**TypeScript Version:** 1.8.4

IDE: VS2015

```ts
`${""}`;
`
    ${""}
}`;//Put cursor here and tap enter to see crash
```

**Expected behavior:** No crash

**Actual behavior:**

```
Error executing script:

Description:
getFormattingEditsAfterKeystroke('...', 29, '
')

Message:


StackTrace:
Error: Debug Failure. False expression:
  at assert(:1676:17)
  at consumeTokenAndAdvanceScanner(:45618:21)
  at processChildNode (:45554:25)
  at processChildNodes (:45601:25)
  at Anonymous function (:45515:21)
...
```

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #7686 
Also took the time to clear up some confusion about `isWhiteSpace` and `isLineBreak`: they are disjoint, but in many places people appear to have assumed that `isWhiteSpace` implies `isLineBreak`. Maybe `isWhiteSpace` should be renamed to `isNonLineBreakWhiteSpace`?

### CODE REVIEW

File: src/compiler/scanner.ts [hunk 1]
```diff
@@ -33,6 +33,7 @@ namespace ts {
         scan(): SyntaxKind;
         // Sets the text for the scanner to scan.  An optional subrange starting point and length
```
Comment (sandersn) [inline src/compiler/scanner.ts]: move new line above old comment

File: src/compiler/scanner.ts [hunk 2]
```diff
@@ -364,6 +365,11 @@ namespace ts {
 
     const hasOwnProperty = Object.prototype.hasOwnProperty;
 
+    export function isWhiteSpaceLike(ch: number): boolean {
+        return isWhiteSpace(ch) || isLineBreak(ch);
+    }
+
+    /** Does not include line breaks. For that, see isWhiteSpaceLike. */
```
Comment (sandersn) [inline src/compiler/scanner.ts]: is there some terminology we could borrow from regexes for this? I know regexes have a distinction between dot-matches-newline and dot-doesn't-match-newline.

File: src/services/formatting/formatting.ts [hunk 3]
```diff
@@ -78,7 +78,7 @@ namespace ts.formatting {
         //  1. the end of the previous line
         //  2. the last non-whitespace character in the current line
         let endOfFormatSpan = getEndLinePosition(line, sourceFile);
```
Comment (sandersn) [inline src/services/formatting/formatting.ts]: The following code looks like it now disagrees with the comment: without `!isLineBreak`, "1. the end of the previous line" will no longer happen.

**Comment by sandersn** [comment]
I don't know enough about the scanner, especially formattingScanner, to know if those changes are right, but otherwise it looks good, with some comments.

**Comment by mhegazy** [comment]
@vladima can you take a look.

File: src/services/formatting/formatting.ts (hunk 3 above)
Comment (andy-ms) [inline src/services/formatting/formatting.ts]: `isWhiteSpace` will already return false on line breaks, so when we reach the end of the previous line the `while` loop will terminate.

File: src/compiler/scanner.ts (hunk 2 above)
Comment (andy-ms) [inline src/compiler/scanner.ts]: The regex terminology for including newlines seems to be "multiline". However, I would normally expect "whitespace" to include newlines. So maybe there should be `isWhiteSpace`, `isWhiteSpaceSingleLine`, and `isLineBreak`?

File: src/services/formatting/formattingScanner.ts [hunk 4]
```diff
@@ -278,5 +280,24 @@ namespace ts.formatting {
             }
             return tokenInfo;
         }
+
+        function skipToEndOf(node: Node): void {
+            scanner.setTextPos(backUpWhitespace());
```
Comment (vladima) [inline src/services/formatting/formattingScanner.ts]: usually `node.end` points to something that is not whitespace. Can you please give an example of a case when end of node points to a whitespace so we need to step back to the first non-whitespace position?

Comment (andy-ms) [inline src/services/formatting/formattingScanner.ts]: I tested and it works without `backUpWhitespace`. Now I'm wondering why I thought it was needed.

**Comment by vladima** [comment]
:+1:

**Pull request merged**